src/lib/shim/src/tls.rs

```diff
@@ -0,0 +1,610 @@
+    }
+}
+
+// The raw byte thread local storage for each thread. Indexes are obtained via
+// `FastThreadIds`.
+#[repr(transparent)]
+struct StoragesType([UnsafeCell<ShimThreadLocalStorage>; TLS_FALLBACK_MAX_THREADS]);
+// SAFETY: normally the UnsafeCell forces this to be !Sync, but we ensure
+// only one thread ever has access to any given UnsafeCell.
+unsafe impl Sync for StoragesType {}
+
+#[cfg(not(miri))]
+mod global_storages {
+    use super::*;
+
+    //  Allocate lazily via `mmap`, to avoid unnecessarily consuming
+    // the memory in processes where we always use native thread local storage.
+    static STORAGES: LazyLock<&'static StoragesType> = LazyLock::const_new(|| {
+        let ptr = unsafe {
+            log::debug!("Allocating fallback thread-local storage");
+            rustix::mm::mmap_anonymous(
+                core::ptr::null_mut(),
+                core::mem::size_of::<StoragesType>(),
+                ProtFlags::READ | ProtFlags::WRITE,
+                MapFlags::PRIVATE,
+            )
+            .unwrap()
+        };
+        unsafe { &*(ptr as *const StoragesType) }
+    });
+
+    pub fn get(idx: usize) -> &'static UnsafeCell<ShimThreadLocalStorage> {
+        &STORAGES.force().0[idx]
+    }
+
+    pub fn initd() -> bool {
+        STORAGES.initd()
+    }
+}
+
+// We can't do dynamic memory allocation via `mmap` under miri, so just
+// use statically allocated storage instead.
+#[cfg(miri)]
+mod global_storages {
+    use super::*;
+
+    static STORAGES: LazyLock<StoragesType> =
+        LazyLock::const_new(|| unsafe { core::mem::zeroed() });
+
+    pub fn get(idx: usize) -> &'static UnsafeCell<ShimThreadLocalStorage> {
+        let rv: &UnsafeCell<ShimThreadLocalStorage> = &STORAGES.force().0[idx];
+        // Unsafely extend the lifetime beyond the LazyLock reference.
+        unsafe { core::mem::transmute(rv) }
+    }
+
+    pub fn initd() -> bool {
+        STORAGES.initd()
+    }
+}
+
+// Should be called by every thread that accesses thread local storage, before
+// it makes the `exit` syscall and after it is done accessing thread local storage.
+//
+// This is a no-op when using native thread-local storage, but lets us reclaim memory
+// otherwise, and is necessary for correctness when `USE_FS_FALLBACK` is enabled.
+//
+// SAFETY: This thread must not access thread local storage again before exiting.
+pub unsafe fn unregister_current_thread() {
+    if !global_storages::initd() {
+        // Nothing to do. Even if another thread happens to be initializing
+        // concurrently, we know the *current* thread isn't registered.
+        return;
+    }
+
+    let current = FastThreadId::current();
+    if let Some(position) = global_thread_ids::get(current) {
+        // Zero for the next thread.
+        // TODO: Consider running drop impls?
+        global_storages::get(position)
+            .get()
+            .write(core::mem::zeroed());
+        global_thread_ids::remove(position);
```

Comment:
It's not possible to safely run a drop impl. Since the thread-locals return `'static` references, you could put one thread-local's `'static` reference in another thread-local. Then if the first thread local is dropped first, it will leave a dangling reference in the other thread-local, which is UB. So I think it should be documented here to never run drop impls on thread-locals.